design/events.md

```diff
@@ -0,0 +1,439 @@
+level in the call stack to intercept events. For example, a commit event could be emitted after the call to
+`org.projectnessie.versioned.VersionStore.commit()`. The delegate pattern is already used at this level to inject
+telemetry, cf. `TracingVersionStore`. A similar delegate pattern could be used for events.
+
+Note: the new storage model has the ability to recover from a failed write upon the next read targeting the same
+reference. Whether the corresponding event should be triggered on the failed write or on the recovery read is left to be
+sorted out later.
+
+### 2.3.1. Asynchronous delivery
+
+**Event delivery implementation must not block a Nessie API call, nor jeopardize the server's stability and
+responsiveness.** Instead, events will require asynchronous processing, and likely, an event emission queue.
+Implementation details will be sorted out later, including mitigation actions when the event queue gets full.
+
+### 2.3.2. Ordering guarantees
+
+The system does not provide any ordering guarantees. Events may be emitted in an out-of-sequence manner.
+
+### 2.3.3. Delivery contract
+
+Event delivery is the act of handing over an event to the subscriber's `EventListener`. For optimal event delivery,
+implementers of the `EventListener` interface are expected to respect the following requirements:
+
+* Methods in the `EventListener` interface SHOULD NOT block the calling thread. If required, implementers of the
+  `EventListener` interface should rather dispatch the calls to separate thread pools, if the underlying event
+  processing requires blocking I/O.
+* Methods in the `EventListener` interface operate in fire and forget mode: the method SHOULD return as soon as the
+  event was published to the destination queue, not when it was read by all consumers. In other words, implementations
+  should not wait for read acknowledgements.
+* The server COULD impose a (configurable) timeout after which the thread calling the `EventListener` method is
+  interrupted and delivery is considered failed.
+* A `RuntimeException` MAY be thrown to indicate that event delivery wasn't possible, e.g. because the underlying
+  pub/sub system is unreachable. But in this case, the exception SHOULD be thrown immediately.
+* For increased resilience, a failed invocation of the subscriber's `EventListener` method MAY be retried a configurable
+  number of times, with exponential backoff.
+
+**This effectively makes the system provide at-most-once delivery guarantees**. Subscribers must be prepared to miss
+some events, and to receive events out of sequence.
+
+Event coalescing (that is, coalescing two or more consecutive, similar events into one) is out of the scope for the MVP
+but could be introduced later as a means to prevent event flooding.
+
+## 2.4. Cross-cutting concerns
+
+### 2.4.1. Security
+
+The initial MVP does not apply authentication nor authorization to event delivery. In other words, each subscriber will
```

Comment:
I think, that's generally fine for this kind of systems. If authZ is required, a separate system should provide that.